jobs/initJobs.ts

```diff
@@ -16,58 +17,77 @@ const RETRY_OPTS = {
 }
 
 const main = async (): Promise<void> => {
-  // sync all db addresses transactions
-  const initTransactionsSync = new Queue('initTransactionsSync', { connection: redisBullMQ })
+  const pricesQueue = new Queue('pricesSync', { connection: redisBullMQ })
+  const initTransactionsQueue = new Queue('initTransactionsSync', { connection: redisBullMQ })
+  const newAddressesQueue = new Queue('newAddressesSync', { connection: redisBullMQ })
+  const subscribeAddressesQueue = new Queue('subscribeAddressesSync', { connection: redisBullMQ })
 
-  // sync current prices
-  const pricesSync = new Queue('pricesSync', { connection: redisBullMQ })
-
-  // try to sync new addresses periodically
-  const newAddressesSync = new Queue('newAddressesSync', { connection: redisBullMQ })
-
-  // create flow
-  const flowProducer = new FlowProducer({ connection: redisBullMQ })
-
-  await flowProducer.add({
-    name: 'syncUnsyncedAddresses',
-    queueName: newAddressesSync.name,
+  const flowJobPrices: FlowJob = {
+    queueName: pricesQueue.name,
+    data: { syncType: 'past' },
+    name: 'syncPastPricesFlow',
+    opts: {
+      removeOnFail: false,
+      jobId: 'syncPastPrices',
+      ...RETRY_OPTS
+    }
+  }
+  const flowJobSyncXECAddresses: FlowJob = {
+    queueName: initTransactionsQueue.name,
+    data: { networkId: XEC_NETWORK_ID },
+    name: 'syncXECAddressesFlow',
+    opts: {
+      removeOnFail: false,
+      jobId: 'syncXECAddresses',
+      ...RETRY_OPTS
+    }
+  }
+  const flowJobSyncBCHAddresses: FlowJob = {
+    queueName: initTransactionsQueue.name,
+    data: { networkId: BCH_NETWORK_ID },
+    name: 'syncBCHAddressesFlow',
+    opts: {
+      removeOnFail: false,
+      jobId: 'syncBCHAddresses',
+      ...RETRY_OPTS
+    }
+  }
+  const flowJobSyncUnsyncedAddresses: FlowJob = {
+    queueName: newAddressesQueue.name,
     data: {},
+    name: 'syncUnsyncedAddressesFlow',
     opts: {
       removeOnComplete: false,
       removeOnFail: { count: 3 },
       jobId: 'syncUnsyncedAddresses',
       ...RETRY_OPTS
-    },
+    }
+  }
+  const flowJobSubscribeAddresses: FlowJob = {
+    queueName: subscribeAddressesQueue.name,
+    data: {},
+    name: 'subscribeAddressesFlow',
+    opts: {
+      removeOnComplete: false,
+      removeOnFail: { count: 3 },
+      jobId: 'subscribeAddresses',
+      ...RETRY_OPTS
+    }
+  }
+
+  const flowProducer = new FlowProducer({ connection: redisBullMQ })
+  await flowProducer.add({
+    ...flowJobSubscribeAddresses,
     children: [
       {
-        name: 'syncBCHAddresses',
-        data: { networkId: BCH_NETWORK_ID },
-        opts: {
-          removeOnFail: false,
-          jobId: 'syncBCHAddresses',
-          ...RETRY_OPTS
-        },
-        queueName: initTransactionsSync.name,
+        ...flowJobSyncUnsyncedAddresses,
```

Comment:
Testing this, I get the following behavior:
![image](https://user-images.githubusercontent.com/21281174/234540664-6f52864f-a2a4-4e1a-88ac-1769b7c1fe7e.png)

The relevant part here is that there is a huge gap between syncing new addresses and subscribing them.

When a new address is added, we have to sync it's previous transactions and subscribe it so we get new transactions too. A gap between this two operations means we potentially lose a transaction that happens in between, so we want to make the gap as small as possible. Better yet, actually, we would first subscribe it and then, right away, we would sync its previous txs because then what we risk is commiting the same tx to the database, which is not a problem if we just parse this error and ignore it.

So I think the order should be inverted in such a way that we only sync the address txs after it is already subscribed.

Reply:
A more basic question: what is the rationale behind separating syncing new addresses and subscribing them into different jobs? As far as I know, those things are very much intertwined so I see no reason for it, while I see the disadvantage of creating the aforementioned gap.

Reply:
Great point, sir. I removed the separate flow for subscribing and now we subscribe the address immediately before syncing it, so there is no gap anymore.